contracts/TwabRewards.sol

```diff
@@ -0,0 +1,422 @@
+        Promotion memory _promotion = _getPromotion(_promotionId);
+        uint256 _numberOfEpochs = _promotion.numberOfEpochs;
+        uint256 _tokensPerEpoch = _promotion.tokensPerEpoch;
+        uint256 _currentEpochId = _getCurrentEpochId(_promotionId);
+
+        if (_currentEpochId == 0) {
+            return _tokensPerEpoch * _numberOfEpochs;
+        } else if (_numberOfEpochs <= _currentEpochId + 1) {
+            return 0;
+        } else {
+            // _tokensPerEpoch * _numberOfEpochsLeft
+            return _tokensPerEpoch * (_numberOfEpochs - (_currentEpochId + 1));
+        }
+    }
+
+    /**
+        @notice Set boolean value for a specific epoch.
+        @dev Bits are stored in a uint256 from right to left.
+        Let's take the example of the following 8 bits word. 0110 0111
+        To set the boolean value to 0 for the epoch id 2, we need to create a mask by shifting 1 to the left by 2 bits.
+        We get: 0000 0001 << 2 = 0000 0100
+        We then OR the mask with the word to set the value.
+        We get: 0110 0111 | 0000 0100 = 0110 0111
+        To set the boolean value to 0 for the epoch id 2, we need to create a mask by shifting 1 to the left by 2 bits and then inverting it.
+        We get: 0000 0001 << 2 = ~(0000 0100) = 1111 1011
+        We then AND the mask with the word to clear the value.
+        We get: 0110 0111 & 1111 1011 = 0110 0011
+        @param _epochs Tightly packed epoch ids with their boolean values
+        @param _epochId Id of the epoch to set the boolean for
+        @param _value Boolean value to set
+        @return Tightly packed epoch ids with the newly boolean value set
+    */
+    function _setClaimedEpoch(
+        uint256 _epochs,
+        uint256 _epochId,
+        bool _value
+    ) public pure returns (uint256) {
+        if (_value) {
+            return _epochs | (uint256(1) << _epochId);
+        } else {
+            return _epochs & ~(uint256(1) << _epochId);
+        }
+    }
+
+    /**
+        @notice Get boolean for a specific epoch id.
+        @dev Bits are stored in a uint256 from right to left.
+        Let's take the example of the following 8 bits word. 0110 0111
+        To retrieve the boolean value for the epoch id 2, we need to shift the word to the right by 2 bits.
+        We get: 0110 0111 >> 2 = 0001 1001
+        We then get the value of the last bit by masking with 1.
+        We get: 0001 1001 & 0000 0001 = 0000 0001 = 1
+        We then return the boolean value true since the last bit is 1.
+        @param _epochs Tightly packed epoch ids with their boolean values
+        @param _epochId Id of the epoch to get the boolean for
+        @return true if the epoch has been claimed, false otherwise
+    */
+    function _getClaimedEpoch(uint256 _epochs, uint256 _epochId) internal pure returns (bool) {
+        uint256 flag = (_epochs >> _epochId) & uint256(1);
+        return (flag == 1 ? true : false);
```

Comment:
think you can just directly return the flag, not sure, try it!

Reply:
actually dont think you can

Reply:
might rename this function to `isClaimedEpoch()` tho to match the external

Reply:
Done: https://github.com/pooltogether/v4-periphery/pull/8/commits/bd258ecd59453e82ceebd0dc03b371c72d14ac56